### ISSUE TITLE
 system.stackoverflowexception nested self referenced class in viewmodel

### ISSUE THREAD

**Issue by mtad2002** [aspnet/AspNetCore#7269]
### Describe the bug
I have a self-referenced class named Category and have another class named Post for my blogging system. I used my Post class in a ViewModel to create a new post. everything worked fine until I migrated my project from asp.net core 2,1 to 2.2 version.

when i want to create new post i get an infinite loop and after that get system.stackoverflowexception.

note that when i use post class as action parameter it works but when I put it in ViewModel it not working. 
 
this is the second bad behavior i got so far since i migrated to 2.2. (See #6446.)

this was the first one that made me rewrite the whole page

### To Reproduce
create a self-reference category class 
use this class in a post class
put that post class in a ViewModel 
try to create a post using ViewModel

### Additional context
PM> dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.2.101
 Commit:    236713b0b7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.101\

Host (useful for support):
  Version: 2.2.0
  Commit:  1249f08fed

.NET Core SDKs installed:
  2.0.0 [C:\Program Files\dotnet\sdk]
  2.0.2 [C:\Program Files\dotnet\sdk]
  2.1.2 [C:\Program Files\dotnet\sdk]
  2.1.4 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.402 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.1.502 [C:\Program Files\dotnet\sdk]
  2.2.101 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

**Comment by pranavkm** [comment]
cc @dougbu

**Comment by dougbu** [comment]
@mtad2002 self-referential model types and other model type loops are not normally problematic. MVC does get into some trouble when the loop includes `IFormFile` properties, involves custom binders and in a few more cases. The issues existed before but are easier to hit in 2.2.x.

Please provide more information about your `Category` and `Post` classes. With that, we should be able to see if this bug is a duplicate and, perhaps, provide a workaround.

**Comment by mtad2002** [comment]
i have IFormfile in both category and post classes , here is my category and post definition 

````
   public class Category : BaseEntity
    {


        public string Name { get; set; }

        public string PersianName { get; set; }

        public string Description { get; set; }

        public string ImagePath { get; set; }

        public virtual ICollection<Post> Posts { get; set; }

        public int? ParentId { get; set; }

        private Category _parent;

        [ForeignKey("ParentId")]
        public virtual Category Parent { get => _parent; set => _parent = value; }

        public virtual ICollection<Category> Children { get; set; }

 public IFormFile Image { get; set; }
      }

 public class Post:BaseEntity
    {
        public Post()
        {
            PostTags = new List<PostTag>();
            Comments = new List<Comment>();
        }

        public string Title { get; set; }

        public string Slug { get; set; }

        public string Abstract { get; set; }

        public string Content { get; set; }

        public string ImagePath { get; set; }

        public IFormFile Image { get; set; }

        [DataType(DataType.Date)]
        public DateTime PubDate { get; set; } = DateTime.UtcNow;

        [DataType(DataType.Date)]
        public DateTime LastModified { get; set; } = DateTime.UtcNow;

        public bool IsPublished { get; set; } = false;

        public int CategoryId { get; set; }

        public Category Category { get; set; }

        public ICollection<PostTag> PostTags { get; set; }

        public ICollection<Comment> Comments { get; set; }
    }

**Comment by dougbu** [comment]
The model seems susceptible to over-binding and should be pared down to a reasonable view model. It currently looks like the entire storage model is exposed. For example, why expose both `Parent` and `ParentId` to the browser?
``` c#
public int? ParentId { get; set; }

private Category _parent; [ForeignKey("ParentId")]

public virtual Category Parent { get => _parent; set => _parent = value; }
```

In any case, the `Category` class as-is duplicates the **#6616** scenario. My PR to fix that (#6793) should be in our 3.0 Preview 3 release or nightlies once we merge the code.

The `Post` class as-is duplicates the **#4802** scenario. #6793 addresses that as well.
